File: src/backend/src/main/java/com/ciet/demo_learn/domain/book/Book.java
package com.ciet.demo_learn.domain.book;

import com.ciet.demo_learn.domain.branch.Branch;
import com.ciet.demo_learn.shared.audit.Auditable;
import com.ciet.demo_learn.shared.uuid.UuidV7;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.FetchType;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "book")
public class Book extends Auditable {

    @Id
    @UuidV7
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Column(name = "title")
    private String title;

    @Column(name = "author")
    private String author;

    @Column(name = "isbn")
    private String isbn;

    @Column(name = "publisher")
    private String publisher;

    @Column(name = "publication_year")
    private Integer publicationYear;

    @Column(name = "genre")
    private String genre;

    @Column(name = "category")
    private String category;

    @Enumerated(EnumType.STRING)
    @Column(name = "condition")
    private BookCondition condition;

    @Column(name = "condition_notes")
    private String conditionNotes;

    @Column(name = "sale_price", precision = 10, scale = 2)
    private BigDecimal salePrice;

    @Column(name = "quantity")
    private Integer quantity;

    @Column(name = "shelf_location")
    private String shelfLocation;

    @Column(name = "registered_at")
    private OffsetDateTime registeredAt;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "branch_id")
    private Branch branch;

    @OneToMany(mappedBy = "book", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<BookImage> images;

    @OneToMany(mappedBy = "book", fetch = FetchType.LAZY)
    private List<BookPriceHistory> priceHistory;

    @OneToOne(mappedBy = "book", fetch = FetchType.LAZY)
    private BookDiscount activeDiscount;

    public UUID getId() { return id; }
    public void setId(UUID id) { this.id = id; }

    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }

    public String getAuthor() { return author; }
    public void setAuthor(String author) { this.author = author; }

    public String getIsbn() { return isbn; }
    public void setIsbn(String isbn) { this.isbn = isbn; }

    public String getPublisher() { return publisher; }
    public void setPublisher(String publisher) { this.publisher = publisher; }

    public Integer getPublicationYear() { return publicationYear; }
    public void setPublicationYear(Integer publicationYear) { this.publicationYear = publicationYear; }

    public String getGenre() { return genre; }
    public void setGenre(String genre) { this.genre = genre; }

    public String getCategory() { return category; }
    public void setCategory(String category) { this.category = category; }

    public BookCondition getCondition() { return condition; }
    public void setCondition(BookCondition condition) { this.condition = condition; }

    public String getConditionNotes() { return conditionNotes; }
    public void setConditionNotes(String conditionNotes) { this.conditionNotes = conditionNotes; }

    public BigDecimal getSalePrice() { return salePrice; }
    public void setSalePrice(BigDecimal salePrice) { this.salePrice = salePrice; }

    public Integer getQuantity() { return quantity; }
    public void setQuantity(Integer quantity) { this.quantity = quantity; }

    public String getShelfLocation() { return shelfLocation; }
    public void setShelfLocation(String shelfLocation) { this.shelfLocation = shelfLocation; }

    public OffsetDateTime getRegisteredAt() { return registeredAt; }
    public void setRegisteredAt(OffsetDateTime registeredAt) { this.registeredAt = registeredAt; }

    public OffsetDateTime getDeletedAt() { return deletedAt; }
    public void setDeletedAt(OffsetDateTime deletedAt) { this.deletedAt = deletedAt; }

    public Branch getBranch() { return branch; }
    public void setBranch(Branch branch) { this.branch = branch; }

    public List<BookImage> getImages() { return images; }
    public void setImages(List<BookImage> images) { this.images = images; }

    public List<BookPriceHistory> getPriceHistory() { return priceHistory; }
    public void setPriceHistory(List<BookPriceHistory> priceHistory) { this.priceHistory = priceHistory; }

    public BookDiscount getActiveDiscount() { return activeDiscount; }
    public void setActiveDiscount(BookDiscount activeDiscount) { this.activeDiscount = activeDiscount; }
}

File: src/frontend/src/types/book.ts
export interface BookImage {
  id: string;
  url: string;
  order: number;
}

export interface Book {
  id: string;
  title: string;
  author: string;
  isbn: string | null;
  publisher: string | null;
  year: number | null;
  category: string;
  condition: 'new' | 'used';
  conditionDescription: string | null;
  salePrice: number;
  description: string | null;
  shelfLocation: string | null;
  branchId: string;
  registeredAt: string;
  active: boolean;
  stockQuantity: number;
  images: BookImage[];
}

export interface BookSummary {
  id: string;
  title: string;
  author: string;
  category: string;
  condition: 'new' | 'used';
  salePrice: number;
  stockQuantity: number;
  shelfLocation: string | null;
}

export interface BookPage {
  content: BookSummary[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
}

export interface IsbnPrefill {
  title: string;
  author: string;
  publisher: string | null;
  year: number | null;
  category: string;
}

export interface BookCreateRequest {
  title: string;
  author: string;
  isbn: string;
  publisher?: string;
  year?: number;
  category: string;
  condition: 'new' | 'used';
  conditionDescription?: string;
  salePrice: number;
  quantity?: number;
  shelfLocation?: string;
  description?: string;
  lotId?: string;
}

export interface BookUpdateRequest {
  title: string;
  author: string;
  isbn: string;
  publisher?: string;
  year?: number;
  category: string;
  conditionDescription?: string;
  salePrice: number;
  quantity?: number;
  shelfLocation?: string;
  description?: string;
}
